README.md


# resm - Resource manager. #

__Authors:__ Averbakh Dmitry ([`adm@ruhub.com`](mailto:adm@ruhub.com)).


### Resm is a simple resource manager with REST API interface and JSON data format

Manager operate with limited resources count, specified in a configuration or environment variables. Resource is a unique rN kind id. Manager provide five operations for resources manipulation: allocate resources for user, deallocate resources, show list of allocated and deallocated resources, show list resources allocated by user and reset allocated resources state.

Resource allocating by username and associated with him up to deallocation. List of allocated resources stored like a tuple list, i.e. [{resource, user}], deallocated resources stored like a simple list. Resource deallocating performed by his unique id. Resources list associated with user you may get by his name.

## Let's get started

### Requirements

[Erlang](http://erlang.org) R16 or higher.

### Installation

#### From sources

Clone the repository:

> git clone git@github.com:hiend/resm.git

Build:

> make

### Run

Production:

> make console

Development:

> make dev-rel console

#### From deb package (debian based linux, e.g. ubuntu)

Download:

> wget https://github.com/hiend/resm/releases/download/1.0.1/resm_1.0.1_amd64.deb

Install:

> sudo dpkg -i resm_1.0.1_amd64.deb

Run:

> sudo service resm start

#### From docker image

Pull image:

> docker pull hiend/resm:1.0.1

Run:

> docker run -dit -p 8008:8008 hiend/resm:1.0.1

### Go to [`http://localhost:8008/list`](http://localhost:8008/list) and check that works!
Stop the running application by typing `halt().` into the REPL console.

----------

# REST API
## 1. Allocate
### GET [/allocate/user](http://localhost:8008/allocate/user)
```curl
RETURNS:
    201, r1
    503, Out of resources
```
## 2. Deallocate
### GET [/deallocate/r1](http://localhost:8008/deallocate/r1)
```curl
RETURNS:
    204
    404, Not allocated
```
## 3. List
### GET [/list](http://localhost:8008/list)
```curl
RETURNS:
    200, {"allocated":[],"deallocated":["r1","r2"]}
    200, {"allocated":{"r1":"user"},"deallocated":["r2"]}
```
## 4. User list
### GET [/list/user](http://localhost:8008/list/user)
```curl
RETURNS:
    200, []
    200, ["r1","r2"]
```
## 5. Reset
### GET [/reset](http://localhost:8008/reset)
```curl
RETURNS:
    204
```

## Modules ##


<table width="100%" border="0" summary="list of modules">
<tr><td><a href="http://github.com/hiend/resm/blob/master/doc/resm.md" class="module">resm</a></td></tr>
<tr><td><a href="http://github.com/hiend/resm/blob/master/doc/resm_manager.md" class="module">resm_manager</a></td></tr>
<tr><td><a href="http://github.com/hiend/resm/blob/master/doc/resm_rest.md" class="module">resm_rest</a></td></tr>
<tr><td><a href="http://github.com/hiend/resm/blob/master/doc/resm_sup.md" class="module">resm_sup</a></td></tr></table>



CHANGELOG.md
# Changelog

## 1.0.1 (2015-06-19)

* Migrate to relang.mk

## 1.0.0 (2015-06-16)

* First public release


test/resm_tests.erl
%%%---------------------------------------------------------------------------------------------------------------------
%%% @author Dmitry Averbakh <adm@ruhub.com>
%%% @copyright 2015, Dmitry Averbakh
%%% @doc
%%% resm resource manager tests.
%%% @end
%%%---------------------------------------------------------------------------------------------------------------------
-module(resm_tests).
-author("Dmitry Averbakh <adm@ruhub.com>").

%% Includes
-include_lib("eunit/include/eunit.hrl").

%% apps for start
-define(APPS, [compiler, syntax_tools, goldrush, crypto, sasl, lager, ranch, cowlib, cowboy, jiffy, resm]).

%% test states
-define(INITIAL_STATE, {ok,{[],[r1,r2,r3,r4,r5,r6,r7,r8,r9,r10]}}).
-define(ALLOCATED_STATE, {ok,{[{r1,user1},{r2,user2}],[r3,r4,r5,r6,r7,r8,r9,r10]}}).
-define(DEALLOCATED_STATE, {ok,{[{r2,user2}], [r3,r4,r5,r6,r7,r8,r9,r10,r1]}}).

%%%=====================================================================================================================
%%% eunit functions
%%%=====================================================================================================================

%%----------------------------------------------------------------------------------------------------------------------
%% @doc
%% Start test environment.
%% @end
%%----------------------------------------------------------------------------------------------------------------------
setup() ->
    [ok = application:start(App) || App <- ?APPS].

%%----------------------------------------------------------------------------------------------------------------------
%% @doc
%% Stop test environment.
%% @end
%%----------------------------------------------------------------------------------------------------------------------
teardown(_) ->
    [ok = application:stop(App) || App <- lists:reverse(?APPS)].

%%%=====================================================================================================================
%%% test functions
%%%=====================================================================================================================

%%----------------------------------------------------------------------------------------------------------------------
%% @doc
%% Common workflow steps.
%% @end
%%----------------------------------------------------------------------------------------------------------------------
resm_test_() ->
    {setup, fun setup/0, fun teardown/1, [
        {"Initial list", fun() ->
            ?assertEqual(?INITIAL_STATE, resm_manager:list())
        end},
        {"Allocate", fun() ->
            ?assertEqual({ok,r1}, resm_manager:allocate(user1)),
            ?assertEqual({ok,r2}, resm_manager:allocate(user2))
        end},
        {"Allocated list", fun() ->
            ?assertEqual(?ALLOCATED_STATE, resm_manager:list())
        end},
        {"Allocated list for user1", fun() ->
            ?assertEqual({ok,[r1]}, resm_manager:list(user1))
        end},
        {"Allocated list for anyuser", fun() ->
            ?assertEqual({ok,[]}, resm_manager:list(anyuser))
        end},
        {"Deallocate r1", fun() ->
            ?assertEqual(ok, resm_manager:deallocate(r1))
        end},
        {"Deallocated list", fun() ->
            ?assertEqual(?DEALLOCATED_STATE, resm_manager:list())
        end},
        {"Reset", fun() ->
            ?assertEqual(ok, resm_manager:reset())
        end},
        {"Reseted list", fun() ->
            ?assertEqual(?INITIAL_STATE, resm_manager:list())
        end},
        {"Out of resources error", fun() ->
            [{ok, _} = resm_manager:allocate(user) || _ <- lists:seq(1, 10)],
            ?assertEqual({error, out_of_resources}, resm_manager:allocate(user))
        end},
        {"Not allocated error", fun() ->
            ?assertEqual({error, not_allocated}, resm_manager:deallocate(not_a_resource))
        end},
        {"Not allocated after reset", fun() ->
            resm_manager:reset(),
            ?assertEqual({error, not_allocated}, resm_manager:deallocate(r1)),
            ?assertEqual({ok, []}, resm_manager:list(user))
        end}
     ]}.

%%%=====================================================================================================================
%%% Internal functions
%%%=====================================================================================================================


test/rest_tests.erl
%%%---------------------------------------------------------------------------------------------------------------------
%%% @author Dmitry Averbakh <adm@ruhub.com>
%%% @copyright 2015, Dmitry Averbakh
%%% @doc
%%% resm rest api handler tests.
%%% @end
%%%---------------------------------------------------------------------------------------------------------------------
-module(rest_tests).
-author("Dmitry Averbakh <adm@ruhub.com>").

%% Includes
-include_lib("eunit/include/eunit.hrl").

%% apps for start
-define(APPS, [compiler, syntax_tools, sasl, lager, ranch, cowlib, cowboy, jiffy, resm, inets]).

%% test states
-define(INITIAL_STATE, "{\"allocated\":{},\"deallocated\":[\"r1\",\"r2\",\"r3\",\"r4\",\"r5\",\"r6\",\"r7\",\"r8\",\"r9\",\"r10\"]}").
-define(ALLOCATED_STATE, "{\"allocated\":{\"r1\":\"user1\",\"r2\":\"user2\"},\"deallocated\":[\"r3\",\"r4\",\"r5\",\"r6\",\"r7\",\"r8\",\"r9\",\"r10\"]}").
-define(DEALLOCATED_STATE, "{\"allocated\":{\"r2\":\"user2\"},\"deallocated\":[\"r3\",\"r4\",\"r5\",\"r6\",\"r7\",\"r8\",\"r9\",\"r10\",\"r1\"]}").

%%%=====================================================================================================================
%%% eunit functions
%%%=====================================================================================================================

%%----------------------------------------------------------------------------------------------------------------------
%% @doc
%% Start test environment.
%% @end
%%----------------------------------------------------------------------------------------------------------------------
setup() ->
    [ok = application:start(App) || App <- ?APPS].

%%----------------------------------------------------------------------------------------------------------------------
%% @doc
%% Stop test environment.
%% @end
%%----------------------------------------------------------------------------------------------------------------------
teardown(_) ->
    [ok = application:stop(App) || App <- lists:reverse(?APPS)].

%%%=====================================================================================================================
%%% test functions
%%%=====================================================================================================================

%%----------------------------------------------------------------------------------------------------------------------
%% @doc
%% Common workflow steps through the rest web api.
%% @end
%%----------------------------------------------------------------------------------------------------------------------
rest_test_() ->
    {setup, fun setup/0, fun teardown/1, [
        {"Initial list", fun() ->
            {ok, {{_HttpVersion, 200, "OK"}, _Headers, ?INITIAL_STATE}} = rest("list")
        end},
        {"Allocate", fun() ->
            {ok, {{_HttpVersion, 201, "Created"}, _Headers, "r1"}} = rest("allocate/user1"),
            {ok, {{_HttpVersion, 201, "Created"}, _Headers, "r2"}} = rest("allocate/user2")
        end},
        {"Allocated list", fun() ->
            {ok, {{_HttpVersion, 200, "OK"}, _Headers, ?ALLOCATED_STATE}} = rest("list")
        end},
        {"Allocated list for user1", fun() ->
            {ok, {{_HttpVersion, 200, "OK"}, _Headers, "[\"r1\"]"}} = rest("list/user1")
        end},
        {"Allocated list for anyuser", fun() ->
            {ok, {{_HttpVersion, 200, "OK"}, _Headers, "[]"}} = rest("list/anyuser")
        end},
        {"Deallocate r1", fun() ->
            {ok, {{_HttpVersion, 204, "No Content"}, _Headers, []}} = rest("deallocate/r1")
        end},
        {"Deallocated list", fun() ->
            {ok, {{_HttpVersion, 200, "OK"}, _Headers, ?DEALLOCATED_STATE}} = rest("list")
        end},
        {"Reset", fun() ->
            {ok, {{_HttpVersion, 204, "No Content"}, _Headers, []}} = rest("reset")
        end},
        {"Reseted list", fun() ->
            {ok, {{_HttpVersion, 200, "OK"}, _Headers, ?INITIAL_STATE}} = rest("list")
        end},
        {"Out of resources error", fun() ->
            [{ok, _} = rest("allocate/user") || _ <- lists:seq(1, 10)],
            {ok, {{_HttpVersion, 503, "Service Unavailable"}, _Headers, "Out of resources"}} = rest("allocate/user")
        end},
        {"Not allocated error", fun() ->
            {ok, {{_HttpVersion, 404, "Not Found"}, _Headers, "Not allocated"}} = rest("deallocate/not_a_resource")
        end},
        {"Not allocated after reset", fun() ->
            {ok, _} = rest("reset"),
            {ok, {{_HttpVersion, 200, "OK"}, _Headers, "[]"}} = rest("list/user")
        end},
        {"Not allocated after reset", fun() ->
            {ok, _} = rest("reset"),
            {ok, {{_HttpVersion, 404, "Not Found"}, _Headers, "Not allocated"}} = rest("deallocate/r1")
        end},
        {"Any request error", fun() ->
            {ok, {{_HttpVersion, 400, "Bad Request"}, _Headers, "Bad request"}} = rest("make/war")
        end}
    ]}.

%%%=====================================================================================================================
%%% Internal functions
%%%=====================================================================================================================

%%----------------------------------------------------------------------------------------------------------------------
%% @doc
%% HTTP request to rest handler.
%% @end
%%----------------------------------------------------------------------------------------------------------------------
rest(Op) when is_list(Op) ->
    httpc:request("http://127.0.0.1:8008/" ++ Op).


test/resm_rest_tests.erl
%%%---------------------------------------------------------------------------------------------------------------------
%%% @author Dmitry Averbakh <adm@ruhub.com>
%%% @copyright 2015, Dmitry Averbakh
%%% @doc
%%% resm rest api handler tests.
%%% @end
%%%---------------------------------------------------------------------------------------------------------------------
-module(resm_rest_tests).
-author("Dmitry Averbakh <adm@ruhub.com>").

%% Includes
-include_lib("eunit/include/eunit.hrl").

%% apps for start
-define(APPS, [compiler, syntax_tools, goldrush, sasl, crypto, lager, ranch, cowlib, cowboy, jiffy, resm, inets]).

%% test states
-define(INITIAL_STATE, "{\"allocated\":{},\"deallocated\":[\"r1\",\"r2\",\"r3\",\"r4\",\"r5\",\"r6\",\"r7\",\"r8\",\"r9\",\"r10\"]}").
-define(ALLOCATED_STATE, "{\"allocated\":{\"r1\":\"user1\",\"r2\":\"user2\"},\"deallocated\":[\"r3\",\"r4\",\"r5\",\"r6\",\"r7\",\"r8\",\"r9\",\"r10\"]}").
-define(DEALLOCATED_STATE, "{\"allocated\":{\"r2\":\"user2\"},\"deallocated\":[\"r3\",\"r4\",\"r5\",\"r6\",\"r7\",\"r8\",\"r9\",\"r10\",\"r1\"]}").

%%%=====================================================================================================================
%%% eunit functions
%%%=====================================================================================================================

%%----------------------------------------------------------------------------------------------------------------------
%% @doc
%% Start test environment.
%% @end
%%----------------------------------------------------------------------------------------------------------------------
setup() ->
    [ok = application:start(App) || App <- ?APPS].

%%----------------------------------------------------------------------------------------------------------------------
%% @doc
%% Stop test environment.
%% @end
%%----------------------------------------------------------------------------------------------------------------------
teardown(_) ->
    [ok = application:stop(App) || App <- lists:reverse(?APPS)].

%%%=====================================================================================================================
%%% test functions
%%%=====================================================================================================================

%%----------------------------------------------------------------------------------------------------------------------
%% @doc
%% Common workflow steps through the rest web api.
%% @end
%%----------------------------------------------------------------------------------------------------------------------
rest_test_() ->
    {setup, fun setup/0, fun teardown/1, [
        {"Initial list", fun() ->
            {ok, {{_HttpVersion, 200, "OK"}, _Headers, ?INITIAL_STATE}} = rest("list")
        end},
        {"Allocate", fun() ->
            {ok, {{_HttpVersion, 201, "Created"}, _Headers, "r1"}} = rest("allocate/user1"),
            {ok, {{_HttpVersion, 201, "Created"}, _Headers, "r2"}} = rest("allocate/user2")
        end},
        {"Allocated list", fun() ->
            {ok, {{_HttpVersion, 200, "OK"}, _Headers, ?ALLOCATED_STATE}} = rest("list")
        end},
        {"Allocated list for user1", fun() ->
            {ok, {{_HttpVersion, 200, "OK"}, _Headers, "[\"r1\"]"}} = rest("list/user1")
        end},
        {"Allocated list for anyuser", fun() ->
            {ok, {{_HttpVersion, 200, "OK"}, _Headers, "[]"}} = rest("list/anyuser")
        end},
        {"Deallocate r1", fun() ->
            {ok, {{_HttpVersion, 204, "No Content"}, _Headers, []}} = rest("deallocate/r1")
        end},
        {"Deallocated list", fun() ->
            {ok, {{_HttpVersion, 200, "OK"}, _Headers, ?DEALLOCATED_STATE}} = rest("list")
        end},
        {"Reset", fun() ->
            {ok, {{_HttpVersion, 204, "No Content"}, _Headers, []}} = rest("reset")
        end},
        {"Reseted list", fun() ->
            {ok, {{_HttpVersion, 200, "OK"}, _Headers, ?INITIAL_STATE}} = rest("list")
        end},
        {"Out of resources error", fun() ->
            [{ok, _} = rest("allocate/user") || _ <- lists:seq(1, 10)],
            {ok, {{_HttpVersion, 503, "Service Unavailable"}, _Headers, "Out of resources"}} = rest("allocate/user")
        end},
        {"Not allocated error", fun() ->
            {ok, {{_HttpVersion, 404, "Not Found"}, _Headers, "Not allocated"}} = rest("deallocate/not_a_resource")
        end},
        {"Not allocated after reset", fun() ->
            {ok, _} = rest("reset"),
            {ok, {{_HttpVersion, 200, "OK"}, _Headers, "[]"}} = rest("list/user")
        end},
        {"Not allocated after reset", fun() ->
            {ok, _} = rest("reset"),
            {ok, {{_HttpVersion, 404, "Not Found"}, _Headers, "Not allocated"}} = rest("deallocate/r1")
        end},
        {"Any request error", fun() ->
            {ok, {{_HttpVersion, 400, "Bad Request"}, _Headers, "Bad request"}} = rest("make/war")
        end}
    ]}.

%%%=====================================================================================================================
%%% Internal functions
%%%=====================================================================================================================

%%----------------------------------------------------------------------------------------------------------------------
%% @doc
%% HTTP request to rest handler.
%% @end
%%----------------------------------------------------------------------------------------------------------------------
rest(Op) when is_list(Op) ->
    httpc:request("http://127.0.0.1:8008/" ++ Op).


src/resm.erl
%%%---------------------------------------------------------------------------------------------------------------------
%%% @author Dmitry Averbakh <adm@ruhub.com>
%%% @copyright 2015, Dmitry Averbakh
%%% @doc
%%% resm application module.
%%% @end
%%%---------------------------------------------------------------------------------------------------------------------
-module(resm).
-author("Dmitry Averbakh <adm@ruhub.com>").

-behaviour(application).

%% Includes
-include("resm.hrl").

%% Application callbacks
-export([start/2, stop/1, get_env/3]).

%%%=====================================================================================================================
%%% Application callbacks
%%%=====================================================================================================================

%%----------------------------------------------------------------------------------------------------------------------
%% @private
%% @doc
%% This function is called whenever an application is started using application:start/[1,2], and should start the
%% processes of the application. If the application is structured according to the OTP design principles as a
%% supervision tree, this means starting the top supervisor of the tree.
%% @end
%%----------------------------------------------------------------------------------------------------------------------

-spec start(StartType :: any(), StartArgs :: any()) -> {'ok', pid()}.

start(_StartType, _StartArgs) ->
    % Routes
    Allocate = {"/allocate/:user", resm_rest, allocate},
    Deallocate = {"/deallocate/:resource", resm_rest, deallocate},
    List = {"/list", resm_rest, list},
    ListUser = {"/list/:user", resm_rest, list_user},
    Reset = {"/reset", resm_rest, reset},
    Any = {'_', resm_rest, any},

    % Cowboy params
    Host = {'_', [Allocate, Deallocate, List, ListUser, Reset, Any]},
    Port = {port, resm:get_env("PORT", port, ?RESM_PORT)},
    Dispatcher = {dispatch, cowboy_router:compile([Host])},
    Env = {env, [Dispatcher]},

    % Start rest API webserver
    {ok, _CPid} = cowboy:start_http(?MODULE, 10, [Port], [Env]),

    % Start supervisor
    {ok, _SPid} = resm_sup:start_link().

%%----------------------------------------------------------------------------------------------------------------------
%% @private
%% @doc
%% This function is called whenever an application has stopped. It is intended to be the opposite of Module:start/2 and
%% should do any necessary cleaning up. The return value is ignored.
%% @end
%%----------------------------------------------------------------------------------------------------------------------

-spec stop(State :: any()) -> 'ok'.

stop(_State) ->
    ok.

%%----------------------------------------------------------------------------------------------------------------------
%% @doc
%% Get environment (system or module) variable.
%% @end
%%----------------------------------------------------------------------------------------------------------------------

-spec get_env(VarName, Key, Default) -> term() when
    VarName :: string(),
    Key     :: atom(),
    Default :: term().

get_env(VarName, Key, Default) when is_list(VarName), is_atom(Key) ->
    case os:getenv(VarName) of
        false ->
            application:get_env(?MODULE, Key, Default);
        Value ->
            Value
    end.

%%%=====================================================================================================================
%%% Internal functions
%%%=====================================================================================================================


src/resm_app.erl
%%%---------------------------------------------------------------------------------------------------------------------
%%% @author Dmitry Averbakh <adm@ruhub.com>
%%% @copyright 2015, Dmitry Averbakh
%%% @doc
%%% resm application module.
%%% @end
%%%---------------------------------------------------------------------------------------------------------------------
-module(resm_app).
-author("Dmitry Averbakh <adm@ruhub.com>").

-behaviour(application).

%% Includes
-include("resm.hrl").

%% Application callbacks
-export([start/2, stop/1]).

%%%=====================================================================================================================
%%% Application callbacks
%%%=====================================================================================================================

%%----------------------------------------------------------------------------------------------------------------------
%% @private
%% @doc
%% This function is called whenever an application is started using application:start/[1,2], and should start the
%% processes of the application. If the application is structured according to the OTP design principles as a
%% supervision tree, this means starting the top supervisor of the tree.
%% @end
%%----------------------------------------------------------------------------------------------------------------------

-spec start(StartType :: any(), StartArgs :: any()) -> {'ok', pid()}.

start(_StartType, _StartArgs) ->
    % Routes
    Allocate = {"/allocate/:user", resm_rest, allocate},
    Deallocate = {"/deallocate/:resource", resm_rest, deallocate},
    List = {"/list", resm_rest, list},
    ListUser = {"/list/:user", resm_rest, list_user},
    Reset = {"/reset", resm_rest, reset},
    Any = {'_', resm_rest, any},

    % Cowboy params
    Host = {'_', [Allocate, Deallocate, List, ListUser, Reset, Any]},
    Port = {port, resm:get_env("PORT", port, ?RESM_PORT)},
    Dispatcher = {dispatch, cowboy_router:compile([Host])},
    Env = {env, [Dispatcher]},

    % Start rest API webserver
    {ok, _CPid} = cowboy:start_http(?MODULE, 10, [Port], [Env]),

    % Start supervisor
    {ok, _SPid} = resm_sup:start_link().

%%----------------------------------------------------------------------------------------------------------------------
%% @private
%% @doc
%% This function is called whenever an application has stopped. It is intended to be the opposite of Module:start/2 and
%% should do any necessary cleaning up. The return value is ignored.
%% @end
%%----------------------------------------------------------------------------------------------------------------------

-spec stop(State :: any()) -> 'ok'.

stop(_State) ->
    ok.

%%%=====================================================================================================================
%%% Internal functions
%%%=====================================================================================================================


src/resm_sup.erl
%%%---------------------------------------------------------------------------------------------------------------------
%%% @author Dmitry Averbakh <adm@ruhub.com>
%%% @copyright 2015, Dmitry Averbakh
%%% @doc
%%% resm supervisor module.
%%% @end
%%%---------------------------------------------------------------------------------------------------------------------
-module(resm_sup).
-author("Dmitry Averbakh <adm@ruhub.com>").

-behaviour(supervisor).

%% API
-export([start_link/0]).

%% Supervisor callbacks
-export([init/1]).

-define(SERVER, ?MODULE).

%%%=====================================================================================================================
%%% API functions
%%%=====================================================================================================================

%%----------------------------------------------------------------------------------------------------------------------
%% @doc
%% Starts the supervisor.
%% @end
%%----------------------------------------------------------------------------------------------------------------------

-spec start_link() -> {'ok', pid()}.

start_link() ->
    {ok, _Pid} = supervisor:start_link({local, ?SERVER}, ?MODULE, []).

%%%=====================================================================================================================
%%% Supervisor callbacks
%%%=====================================================================================================================

%%----------------------------------------------------------------------------------------------------------------------
%% @private
%% @doc
%% Whenever a supervisor is started using supervisor:start_link/[2,3], this function is called by the new process to
%% find out about restart strategy, maximum restart frequency and child specifications.
%% @end
%%----------------------------------------------------------------------------------------------------------------------

-spec init(Args) -> {ok, {{Strategy, MaxR, MaxT}, [Spec]}} when
    Args     :: any(),
    Strategy :: supervisor:strategy(),
    MaxR     :: non_neg_integer(),
    MaxT     :: non_neg_integer(),
    Spec     :: supervisor:child_spec().

init(_Args) ->
    RestartStrategy = {one_for_one, 5, 10},
    ChildSpec = {resm_manager, {resm_manager, start_link, []}, permanent, 2000, worker, [resm_manager]},

    {ok, {RestartStrategy, [ChildSpec]}}.

%%%=====================================================================================================================
%%% Internal functions
%%%=====================================================================================================================


src/resm_rest.erl
%%%---------------------------------------------------------------------------------------------------------------------
%%% @author Dmitry Averbakh <adm@ruhub.com>
%%% @copyright 2015, Dmitry Averbakh
%%% @doc
%%% resm http rest api handler module.
%%% @end
%%%---------------------------------------------------------------------------------------------------------------------
-module(resm_rest).
-author("Dmitry Averbakh <adm@ruhub.com>").

-behavior(cowboy_http_handler).

%% Includes
-include("resm.hrl").

%% cowboy callbacks
-export([init/3, handle/2, terminate/3]).

%% response headers
-define(CONTENT_ENCODING, {<<"content-encoding">>, <<"utf-8">>}).
-define(CONTENT_TYPE, {<<"content-type">>, <<"application/json">>}).

%%%=====================================================================================================================
%%% cowboy callbacks
%%%=====================================================================================================================

%%----------------------------------------------------------------------------------------------------------------------
%% @private
%% @doc
%% This function is meant for initialization. It receives information about the transport and protocol used, along
%% with the handler options from the dispatch list, and allows you to upgrade the protocol if needed.
%% @end
%%----------------------------------------------------------------------------------------------------------------------

-spec init({tcp, http}, Req, State) -> {'ok', Req, State} when
    Req   :: cowboy_req:req(),
    State :: any().

init({tcp, http}, Req, State) ->
    {ok, Req, State}.

%%----------------------------------------------------------------------------------------------------------------------
%% @private
%% @doc
%% This function for handling the request. It receives the request and the state.
%% @end
%%----------------------------------------------------------------------------------------------------------------------

-spec handle(Req, State) -> {'ok', Req, State} when
    Req   :: cowboy_req:req(),
    State :: atom().

handle(Req, allocate) ->
    {User, Req1} = cowboy_req:binding(user, Req),
    {ok, Replied} = case resm_manager:allocate(binary_to_atom(User, utf8)) of
        {ok, Resource} ->
            cowboy_req:reply(201, [?CONTENT_ENCODING, ?CONTENT_TYPE], atom_to_binary(Resource, utf8), Req1);
        {error, out_of_resources} ->
            cowboy_req:reply(503, [], <<"Out of resources">>, Req1)
    end,
    {ok, Replied, allocate};

handle(Req, deallocate) ->
    {Resource, Req1} = cowboy_req:binding(resource, Req),
    {ok, Replied} = case resm_manager:deallocate(binary_to_atom(Resource, utf8)) of
        ok ->
            {ok, Req1};
        {error, not_allocated} ->
            cowboy_req:reply(404, [], <<"Not allocated">>, Req1)
    end,
    {ok, Replied, deallocate};

handle(Req, list) ->
    {ok, {Allocated, Deallocated}} = resm_manager:list(),
    EJSON = {[{allocated, {Allocated}}, {deallocated, Deallocated}]},
    {ok, Replied} = cowboy_req:reply(200, [?CONTENT_ENCODING, ?CONTENT_TYPE], jiffy:encode(EJSON), Req),
    {ok, Replied, list};

handle(Req, list_user) ->
    {User, Req1} = cowboy_req:binding(user, Req),
    {ok, Resources} = resm_manager:list(binary_to_atom(User, utf8)),
    {ok, Replied} = cowboy_req:reply(200, [?CONTENT_ENCODING, ?CONTENT_TYPE], jiffy:encode(Resources), Req1),
    {ok, Replied, list_user};

handle(Req, reset) ->
    ok = resm_manager:reset(),
    {ok, Req, reset};

handle(Req, any) ->
    {ok, Replied} = cowboy_req:reply(400, [], <<"Bad request">>, Req),
    {ok, Replied, any}.

%%----------------------------------------------------------------------------------------------------------------------
%% @private
%% @doc
%% This function for cleaning up. It also receives the request and the state.
%% @end
%%----------------------------------------------------------------------------------------------------------------------

-spec terminate(Reason :: any(), Req :: any(), State :: any()) -> 'ok'.

terminate(_Reason, _Req, _State) ->
    ok.

%%%=====================================================================================================================
%%% Internal functions
%%%=====================================================================================================================
